### ISSUE TITLE
Possible bug in bot mitigation

### ISSUE THREAD

**Issue by turban1988** [mozilla/OpenWPM#187]
Hi,

When I am using openWPM the crawl sometimes crashes with the following exception:
```

25/05/2018 11:44:32 | BROWSER 40: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/home/openwpm/3rd_party_sharing/data_collection/automation/BrowserManager.py", line 405, in BrowserManager
    browser_params, manager_params, extension_socket)
  File "/home/openwpm/3rd_party_sharing/data_collection/automation/Commands/command_executor.py", line 15, in execute_command
    extension_socket=extension_socket)
  File "/home/openwpm/3rd_party_sharing/data_collection/automation/Commands/browser_commands.py", line 146, in get_website
    bot_mitigation(webdriver)
  File "/home/openwpm/3rd_party_sharing/data_collection/automation/Commands/browser_commands.py", line 59, in bot_mitigation
    action.perform()
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/common/action_chains.py", line 80, in perform
    self.w3c_actions.perform()
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/common/actions/action_builder.py", line 76, in perform
    self.driver.execute(Command.W3C_ACTIONS, enc)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 308, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 194, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: Parameter 'y' was not a positive integer
```

I looked at the code throwing the exeption:
https://github.com/citp/OpenWPM/blob/277b7a375d0c72d4f4b0ba9992dd882fa56b8388/automation/Commands/browser_commands.py#L50-L56

It seems that during the bot mitigation process a negative random number is generated (Line 53) which cannot be processed (Line 56).

I fixed the issue by not allowing negativ random numbers:
```
                x = random.randint(0, move_max)
                y = random.randint(0, move_max)
```

Did anyone else experience this issue, or am I missing something?

Cheers,
Tobias

**Comment by englehardt** [comment]
Thanks for the report! I have a few questions:
1. Did you make any changes to the default window size?
2. Does the crash happen for all negative values? Or only larger negative values. If it's all values, Selenium may have changed the default pointer location to `(0,0)`

**Comment by turban1988** [comment]
Hi,

I set the window size to 800x600 (`profile_settings['screen_res'] = (800,600)`).

I ran two simple tests. I altered Line 53 to  : `y = -10000`  and  `y= -1`. I visited a random Youtube page and both runs crashed.

```
8/05/2018 09:52:03 | BROWSER 6: driver: 1527493923810	addons.xpi	DEBUG	Updating XPIState for {"id":"@openwpm","syncGUID":"{a8a2daed-0445-45d6-9a46-fa0f1b74bda0}","location":"app-profile","version":"0.0.1","type":"extension","internalName":null,"updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"openwpm","description":"Extension with socket interface into OpenWPM","creator":"Steven Englehardt","homepageURL":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/tmp/rust_mozprofile.UZc7sU43AGJD/extensions/@openwpm","installDate":1527493923000,"updateDate":1527493923000,"applyBackgroundUpdates":1,"bootstrap":true,"skinnable":false,"size":94227,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":true,"hasBinaryComponents":false,"strictCompatibility":false,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"38.0a1","maxVersion":"*"}],"targetPlatforms":[],"multiprocessCompatible":false,"signedState":0,"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false,"mpcOptedOut":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923810	addons.xpi	DEBUG	getModTime: Recursive scan of @openwpm
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923812	addons.xpi	DEBUG	Updating XPIState for {"id":"firefox@getpocket.com","syncGUID":"{4eb17ab6-c4b8-4c0b-ae19-8ef65e0df7b5}","location":"app-system-defaults","version":"1.0.5","type":"extension","internalName":null,"updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Pocket","description":"When you find something you want to view later, put it in Pocket.","creator":null,"homepageURL":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/firefox@getpocket.com.xpi","installDate":1525127736000,"updateDate":1525127736000,"applyBackgroundUpdates":1,"bootstrap":true,"skinnable":false,"size":920281,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":false,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"52.8.0","maxVersion":"52.*"}],"targetPlatforms":[],"multiprocessCompatible":true,"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false,"mpcOptedOut":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923812	addons.xpi	DEBUG	getModTime: Recursive scan of firefox@getpocket.com
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923814	addons.xpi	DEBUG	Updating XPIState for {"id":"webcompat@mozilla.org","syncGUID":"{699037ca-5af4-423f-a365-b8e226a97636}","location":"app-system-defaults","version":"1.0","type":"extension","internalName":null,"updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Web Compat","description":"Urgent post-release fixes for web compatibility.","creator":null,"homepageURL":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/webcompat@mozilla.org.xpi","installDate":1525127736000,"updateDate":1525127736000,"applyBackgroundUpdates":1,"bootstrap":true,"skinnable":false,"size":1456,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":false,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"52.8.0","maxVersion":"52.*"}],"targetPlatforms":[],"multiprocessCompatible":true,"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false,"mpcOptedOut":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923814	addons.xpi	DEBUG	getModTime: Recursive scan of webcompat@mozilla.org
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923814	addons.xpi	DEBUG	Updating XPIState for {"id":"e10srollout@mozilla.org","syncGUID":"{15e88421-7212-471f-bde1-6103efbfface}","location":"app-system-defaults","version":"1.10","type":"extension","internalName":null,"updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Multi-process staged rollout","description":"Staged rollout of Firefox multi-process feature.","creator":null,"homepageURL":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/e10srollout@mozilla.org.xpi","installDate":1525127736000,"updateDate":1525127736000,"applyBackgroundUpdates":1,"bootstrap":true,"skinnable":false,"size":8145,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":false,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"52.8.0","maxVersion":"52.*"}],"targetPlatforms":[],"multiprocessCompatible":true,"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false,"mpcOptedOut":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923814	addons.xpi	DEBUG	getModTime: Recursive scan of e10srollout@mozilla.org
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923814	addons.xpi	DEBUG	Updating XPIState for {"id":"aushelper@mozilla.org","syncGUID":"{51758df9-1889-4c25-b2b0-ec5b4c0492e7}","location":"app-system-defaults","version":"2.0","type":"extension","internalName":null,"updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Application Update Service Helper","description":"Sets value(s) in the update url based on custom checks.","creator":null,"homepageURL":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/aushelper@mozilla.org.xpi","installDate":1525127736000,"updateDate":1525127736000,"applyBackgroundUpdates":1,"bootstrap":true,"skinnable":false,"size":8488,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":false,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"52.8.0","maxVersion":"52.*"}],"targetPlatforms":[],"multiprocessCompatible":true,"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false,"mpcOptedOut":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi	DEBUG	getModTime: Recursive scan of aushelper@mozilla.org
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi	DEBUG	Updating XPIState for {"id":"{972ce4c6-7e08-4474-a285-3208198ce6fd}","syncGUID":"{399aed6a-34df-4e8f-8940-d3357fd944aa}","location":"app-global","version":"52.8.0","type":"theme","internalName":"classic/1.0","updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{"32":"icon.png","48":"icon.png"},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Default","description":"The default theme.","creator":"Mozilla","homepageURL":null,"contributors":["Mozilla Contributors"]},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/extensions/{972ce4c6-7e08-4474-a285-3208198ce6fd}.xpi","installDate":1525127736000,"updateDate":1525127736000,"applyBackgroundUpdates":1,"skinnable":true,"size":4932,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":true,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"52.8.0","maxVersion":"52.8.0"}],"targetPlatforms":[],"seen":true,"dependencies":[],"hasEmbeddedWebExtension":false}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi	DEBUG	getModTime: Recursive scan of {972ce4c6-7e08-4474-a285-3208198ce6fd}
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi	DEBUG	Add-on @openwpm blocks e10s rollout.
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	DeferredSave.extensions.json	DEBUG	Save changes
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi	DEBUG	Updating database with changes to installed add-ons
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923816	addons.xpi-utils	DEBUG	Updating add-on states
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923818	addons.xpi-utils	DEBUG	Writing add-ons list
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923820	addons.xpi	DEBUG	Registering manifest for /tmp/rust_mozprofile.UZc7sU43AGJD/extensions/@openwpm
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923826	addons.xpi	DEBUG	Calling bootstrap method startup on @openwpm version 0.0.1
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923872	addons.xpi	DEBUG	Registering manifest for /home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/aushelper@mozilla.org.xpi
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923876	addons.xpi	DEBUG	Calling bootstrap method startup on aushelper@mozilla.org version 2.0
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923876	addons.xpi	DEBUG	Registering manifest for /home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/e10srollout@mozilla.org.xpi
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923878	addons.xpi	DEBUG	Calling bootstrap method startup on e10srollout@mozilla.org version 1.10
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923878	addons.xpi	DEBUG	Registering manifest for /home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/firefox@getpocket.com.xpi
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923880	addons.xpi	DEBUG	Calling bootstrap method startup on firefox@getpocket.com version 1.0.5
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923880	addons.xpi	DEBUG	Registering manifest for /home/openwpm/OpenWPM-master/data_collection/firefox-bin/browser/features/webcompat@mozilla.org.xpi
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923886	addons.xpi	DEBUG	Calling bootstrap method startup on webcompat@mozilla.org version 1.0
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923890	addons.manager	DEBUG	Registering shutdown blocker for XPIProvider
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923892	addons.manager	DEBUG	Provider finished startup: XPIProvider
28/05/2018 09:52:03 | BROWSER 6: driver: 1527493923892	addons.manager	DEBUG	Starting provider: LightweightThemeManager
28/05/2018 09:52:05 | Cookie instrumentation enabled
28/05/2018 09:52:05 | HTTP Instrumentation enabled
BrowserManager       - DEBUG    - BROWSER 6: Looking for extension port information in /tmp/rust_mozprofile.UZc7sU43AGJD/
28/05/2018 09:52:06 | BROWSER 6: Looking for extension port information in /tmp/rust_mozprofile.UZc7sU43AGJD/
BrowserManager       - DEBUG    - BROWSER 6: Connecting to extension on port 33175
28/05/2018 09:52:06 | BROWSER 6: Connecting to extension on port 33175
BrowserManager       - DEBUG    - BROWSER 6: BrowserManager ready.
28/05/2018 09:52:06 | BROWSER 6: BrowserManager ready.
BrowserManager       - DEBUG    - BROWSER 6: Browser spawn sucessful!
28/05/2018 09:52:06 | BROWSER 6: Browser spawn sucessful!
28/05/2018 09:52:06 | BROWSER 6: Looking for extension port information in /tmp/rust_mozprofile.UZc7sU43AGJD/
28/05/2018 09:52:06 | BROWSER 6: Connecting to extension on port 33175
28/05/2018 09:52:06 | BROWSER 6: BrowserManager ready.
28/05/2018 09:52:06 | BROWSER 6: Browser spawn sucessful!
TaskManager          - DEBUG    - Checking command failure status indicator...
28/05/2018 09:52:06 | Checking command failure status indicator...
BrowserManager       - DEBUG    - BROWSER 6: Joining command thread
28/05/2018 09:52:06 | BROWSER 6: Joining command thread
BrowserManager       - INFO     - BROWSER 6: EXECUTING COMMAND: ('GET', 'https://www.youtube.com/watch?v=IR3pQend4UE', 0, 10)
28/05/2018 09:52:06 | BROWSER 6: EXECUTING COMMAND: ('GET', 'https://www.youtube.com/watch?v=IR3pQend4UE', 0, 10)
28/05/2018 09:52:06 | Checking command failure status indicator...
28/05/2018 09:52:06 | BROWSER 6: Joining command thread
28/05/2018 09:52:06 | BROWSER 6: EXECUTING COMMAND: ('GET', 'https://www.youtube.com/watch?v=IR3pQend4UE', 0, 10)
28/05/2018 09:52:06 | Visit Id: 10
BrowserManager       - INFO     - BROWSER 6: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/home/openwpm/OpenWPM-master/data_collection/automation/BrowserManager.py", line 405, in BrowserManager
    browser_params, manager_params, extension_socket)
  File "/home/openwpm/OpenWPM-master/data_collection/automation/Commands/command_executor.py", line 15, in execute_command
    extension_socket=extension_socket)
  File "/home/openwpm/OpenWPM-master/data_collection/automation/Commands/browser_commands.py", line 147, in get_website
    bot_mitigation(webdriver)
  File "/home/openwpm/OpenWPM-master/data_collection/automation/Commands/browser_commands.py", line 60, in bot_mitigation
    action.perform()
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/common/action_chains.py", line 80, in perform
    self.w3c_actions.perform()
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/common/actions/action_builder.py", line 76, in perform
    self.driver.execute(Command.W3C_ACTIONS, enc)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 308, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 194, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: Parameter 'y' was not a positive integer


```

**Comment by englehardt** [comment]
Thanks for the update!

It looks like this is a known issue that was fixed in geckodriver 0.16.0 (see https://github.com/mozilla/geckodriver/issues/649#issuecomment-294834777). That means we're blocked on the upgrade to ESR 60 (see #148).

**Comment by englehardt** [comment]
Now that we've upgraded past geckodriver v0.16.0 this should be fixed.
